[Rahul Agarwal] iP

rahula1008 opened the pull request:
# SirPotato
> "One machine can do the work of fifty ordinary men.  No machine can do the work of one extraordinary man." - Elbert Hubbard [Source](https://www.forbes.com/sites/robertszczerba/2015/02/09/20-great-technology-quotes-to-inspire-amaze-and-amuse/)

SirPotato is an easy chatbot that acts as your personal assistant and remembers all of your information for you. It's
- text-based GUI
- easily learnable
- ~quick~ VERY QUICK to use

The only thing you need to do is,
1. download the app from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)
2. run the app by double-clicking it
3. add whatever tasks you need the bot to store
4. rest easy and let the bot handle your task management :smile:

The best part about all this is that it's **free!**

Features:

- [X] Task management 
- [ ] Deadline management (coming soon)
- [ ] Reminders (coming soon)

If you are a Java coder, you can even use it to practice coding. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

agarwalrahul1008 committed:
Level 0: Rename, Greet, Exit.

agarwalrahul1008 committed:
Level 1: Echo

agarwalrahul1008 committed:
Level 2: Add, List

agarwalrahul1008 committed:
Level 3: Mark as done

agarwalrahul1008 committed:
Level 4: Todos, Deadlines, Events

agarwalrahul1008 committed:
Remove .class files from tracking

agarwalrahul1008 committed:
Automated Text UI Testing

agarwalrahul1008 committed:
Level 5: Exceptions/Errors

agarwalrahul1008 committed:
Level-6: Delete

agarwalrahul1008 committed:
Level 7: Save

agarwalrahul1008 committed:
Merge branch-Level-7 into master

agarwalrahul1008 committed:
Level 8: Dates and Times

agarwalrahul1008 committed:
Merge branch-Level-8 into master

agarwalrahul1008 committed:
Use more OOP

agarwalrahul1008 committed:
Merge More-OOP Branch into master

agarwalrahul1008 committed:
Organise into packages

agarwalrahul1008 committed:
Merge Packages branch into master

agarwalrahul1008 committed:
Merged Gradle into master

agarwalrahul1008 committed:
Create JUnit tests

agarwalrahul1008 committed:
Merge JUnit Test branch into master

agarwalrahul1008 committed:
Create the jar file

agarwalrahul1008 committed:
Add JavaDocs

agarwalrahul1008 committed:
Follow coding standard

agarwalrahul1008 committed:
Merge JavaDoc branch into master

agarwalrahul1008 committed:
Merge coding standard branch

agarwalrahul1008 committed:
Implement find feature

agarwalrahul1008 committed:
Merge branch level 9 into master

Jaynon commented on src/main/java/sirpotato/AddCommand.java:
```diff
@@ -0,0 +1,21 @@
+package sirpotato;
```
It would be nice if the command files are placed in a separate sub-folder in the package

rahula1008 replied:
Yes, I would agree with that comment. As I didn't have enough time then, I just did the minimal level so I could move on to the next task. Thank you for your input!

Jaynon commented on src/main/java/sirpotato/MarkCommand.java:
```diff
@@ -0,0 +1,16 @@
+
+public class MarkCommand extends Command {
+
+    private int itemNumber;
```
maybe `index` could be a better name since the items are in a list?

rahula1008 replied:
Yes, good observation, thanks for the input!

Jaynon commented on src/main/java/sirpotato/Parser.java:
```diff
@@ -0,0 +1,111 @@
+     * Validates user input
+     * @param userInput The user's input to the chatbot
+     */
+    public static void checkForErrors(String userInput) throws DukeException {
+        if (userInput.startsWith("todo")) {
+            if (userInput.length() <= 5 || userInput.substring(5).isEmpty()) {
+                throw new DukeException("Mate, you have got to give us a description of the task");
+            }
+        } else if (userInput.startsWith("deadline")) {
+            String[] sectionedString = userInput.split("/by ");
+            if (sectionedString.length < 2 || 
+                sectionedString[0].substring(9).isEmpty() || 
+                sectionedString[1].isEmpty()) {
+                throw new DukeException("Mate, a deadline must include a task, and the deadline");
+            }
+            validateDateFormat(sectionedString[1].trim());
+        } else if (userInput.startsWith("event")) {
+            String[] sectionedString = userInput.split(" /from | /to ");
+            if (sectionedString.length < 3 || sectionedString[0].substring(6).isEmpty() ||
+                sectionedString[1].isEmpty() || sectionedString[2].isEmpty()) {
+                throw new DukeException("Mate, an event should have the description, the start, and the end.");
+            }
+            validateDateFormat(sectionedString[1].trim());
+            validateDateFormat(sectionedString[2].trim());
+        } else if (userInput.startsWith("delete")) {
+            if (userInput.length() <= 7) {
+                throw new DukeException("You need to say which item to delete");
+            }
+        } else if (!userInput.equals("bye") && !userInput.equals("list") && 
+                   !userInput.startsWith("mark") && !userInput.startsWith("unmark")
+                   && !userInput.startsWith("find")) {
+            throw new DukeException("I'm sorry, that is not a valid input");
+        }
```
It is good that the method's name is a noun, although for the method body it would be great if there are some whitespaces between if-else statements to improve readability

rahula1008 replied:
Thanks for the suggestion - that's a good point!

Jaynon commented on src/test/java/sirpotato/ParserTest.java:
```diff
@@ -0,0 +1,48 @@
+public class ParserTest {
+
+    @Test
+    public void parseCommand_validTodoCommand_returnsAddCommand() throws DukeException {
```
names for all test methods looks good

rahula1008 replied:
Thanks! 😄

Jaynon reviewed (commented):
Quality of names looks good! although it would be nice if the codefiles are partitioned into their own packages, such as command, tasks, storage etc.

ngkhengyang commented on src/main/java/sirpotato/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+ * such as forgetting the description of the task
+ * or forgetting to put the /by
+ */
+public class DukeException extends Exception {
```
The chatbot is no longer named Duke, maybe it would make more sense to rename the exception to `SirPotatoException` or even something more technical like `InvalidSyntaxException`?

ngkhengyang commented on src/main/java/sirpotato/Parser.java:
```diff
@@ -0,0 +1,111 @@
+
+    /**
+     * Validates the date that user inputs
+     * If not in dd-MM-yyyy format, will throw an exception
```
Shouldn't there be an empty line between the description and parameters section in javadoc?
I've noticed that this occurs for other Javadoc comments as well.

ngkhengyang commented on src/main/java/sirpotato/SirPotato.java:
```diff
@@ -0,0 +1,71 @@
+import java.time.LocalDate;
+
+import java.util.ArrayList; 
+import java.util.Scanner;  
```
I like how you left an empty line between each group of imports and keep the order of imports consistent so it's easier to browse the list.

ngkhengyang commented on src/main/java/sirpotato/Storage.java:
```diff
@@ -0,0 +1,106 @@
+     * Converts date in text file to LocalDate object 
+     * @return LocalDate object in the dd-MM-yyyy format
+     */
+    private LocalDate parseData(String dateToParse) throws DateTimeParseException {
```
Since this method focuses on parsing date, maybe the method can be named `parseDate` so it's more specific?

ngkhengyang reviewed (commented):
While there are some minor nits, I think overall you did a good job in upholding the coding standard so keep it up!

agarwalrahul1008 committed:
Implement checkstyle

agarwalrahul1008 committed:
Merge branch-A-CheckStyle into master

agarwalrahul1008 committed:
Add GUI using JavaFX

agarwalrahul1008 committed:
Merge level 10 branch into master

agarwalrahul1008 committed:
Fix Code Quality

rahula1008 committed:
Merge pull request #1 from rahula1008/branch-A-CodeQuality

Merge code quality branch into master

agarwalrahul1008 committed:
Add assertions

rahula1008 committed:
Merge pull request #2 from rahula1008/branch-A-Assertions

Merge assertions branch into master

agarwalrahul1008 committed:
Add C-Sort Extension

agarwalrahul1008 committed:
Merge C-Sort branch into master

agarwalrahul1008 committed:
Remove horizontal line and indents from UI

The UI class displays horizontal lines and indents from when it was command-line based

As a step towards updating the UI to fit the new GUI, let's change the display methods to just display the necesary text instead of the horizontal lines and indents.

agarwalrahul1008 committed:
Add streams

agarwalrahul1008 committed:
Add title to GUI window

agarwalrahul1008 committed:
Improve GUI

agarwalrahul1008 committed:
Fix file creation in Storage class

agarwalrahul1008 committed:
Merge Fix-Storage branch into master

agarwalrahul1008 committed:
Delete data/list.txt file

agarwalrahul1008 committed:
Add user guide

agarwalrahul1008 committed:
Improve README.md

agarwalrahul1008 committed:
Improve User Guide

agarwalrahul1008 committed:
Improve user guide

agarwalrahul1008 committed:
Improve User Guide

agarwalrahul1008 committed:
docs/README.md: Add FAQ section

agarwalrahul1008 committed:
Fix UI findTasks method

agarwalrahul1008 committed:
docs/README.md: Fix typos

agarwalrahul1008 committed:
Fix mark / unmark command bug

agarwalrahul1008 committed:
Improve input validation

agarwalrahul1008 committed:
Add checks to input validations

agarwalrahul1008 committed:
docs/README.md: Note on white space

agarwalrahul1008 committed:
docs/README.md: Add note on whitespace

agarwalrahul1008 committed:
docs/README.md: Fix formatting

agarwalrahul1008 committed:
docs/README.md: Update releases link

agarwalrahul1008 committed:
docs/README.md: Update Deadline description

agarwalrahul1008 committed:
Add input validation checks

agarwalrahul1008 committed:
Shorten parsing methods

agarwalrahul1008 committed:
Update parser & relevant documentation

agarwalrahul1008 committed:
Change DukeException to SirPotato exception

agarwalrahul1008 committed:
Move all java files into sirpotato package

agarwalrahul1008 committed:
Remove dead code

agarwalrahul1008 committed:
Fix minor code quality changes
